textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -0,0 +1,167 @@
+package edu.uci.ics.textdb.dataflow.regexmatch;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import edu.uci.ics.textdb.common.constants.DataConstants;
+
+
+public class GramBooleanQuery {
+	public enum QueryOp {
+		NONE, // doesn't match any string
+		ANY,  // matches any string
+		
+		AND,
+		OR
+	}
+	QueryOp operator;
+	List<String> operandList;
+	List<GramBooleanQuery> subQueryList;
+	
+	private static int gramNum = 3;
+	
+
+	public GramBooleanQuery(QueryOp operator) {
+		this.operator = operator;
+		operandList = new ArrayList<String>();
+		subQueryList = new ArrayList<GramBooleanQuery>();
+	}
+	
+	public boolean equals(GramBooleanQuery query) {
+		if (this.operator != query.operator
+			|| this.operandList.size() != query.operandList.size()
+			|| this.subQueryList.size() != query.subQueryList.size()) {
+			return false;
+		}
+		
+		Set<String> operandSet = new HashSet<String>(this.operandList);
+		if (!operandSet.equals(new HashSet<String>(query.operandList))) {
+			return false;
+		}
+		
+		if (this.subQueryList.size() == 0) {
```

Comment:
I wonder why we cannot use the same hashset-baseed method to compare two sets of subqueries, as long as we have a good  `hashCode()` function?  If so, isn't that much simpler, since HashSet will do the bucketization and recursive call automatically.